### PR TITLE
fix: correctly check for properties in thaw/snow

### LINKED ISSUE #2963: Out of bounds exception?
### Server Implementation

Paper

### Server Version

1.21

### Describe the bug

```
[20:24:32 ERROR]: An unexpected error while handling a FastAsyncWorldEdit command
java.lang.IndexOutOfBoundsException: Index 3 out of bounds for length 3
        at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100) ~[?:?]
        at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106) ~[?:?]
        at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302) ~[?:?]
        at java.base/java.util.Objects.checkIndex(Objects.java:385) ~[?:?]
        at java.base/java.util.ArrayList.get(ArrayList.java:427) ~[?:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.registry.state.AbstractProperty.getValue(AbstractProperty.java:98) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.world.block.BlockState.getState(BlockState.java:322) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.function.block.SnowSimulator.apply(SnowSimulator.java:124) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.function.visitor.LayerVisitor.resume(LayerVisitor.java:152) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.function.operation.Operations.completeLegacy(Operations.java:55) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.EditSession.simulateSnow(EditSession.java:2817) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.command.UtilityCommands.snow(UtilityCommands.java:565) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.command.UtilityCommandsRegistration.cmd$snow(UtilityCommandsRegistration.java:670) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/org.enginehub.piston.CommandManager.execute(CommandManager.java:157) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.extension.platform.PlatformCommandManager.lambda$handleCommandOnCurrentThread$16(PlatformCommandManager.java:712) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.extension.platform.PlatformCommandManager.handleCommandTask(PlatformCommandManager.java:733) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.extension.platform.PlatformCommandManager.handleCommandOnCurrentThread(PlatformCommandManager.java:714) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.extension.platform.PlatformCommandManager.lambda$handleCommand$14(PlatformCommandManager.java:685) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.sk89q.worldedit.extension.platform.AbstractPlayerActor.lambda$runAction$1(AbstractPlayerActor.java:672) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.fastasyncworldedit.core.util.task.AsyncNotifyKeyedQueue.lambda$run$0(AsyncNotifyKeyedQueue.java:48) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.fastasyncworldedit.core.util.task.AsyncNotifyKeyedQueue.lambda$call$1(AsyncNotifyKeyedQueue.java:58) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar/com.fastasyncworldedit.core.util.task.KeyQueuedExecutorService$KeyRunner.lambda$run$0(KeyQueuedExecutorService.java:150) ~[FastAsyncWorldEdit-Bukkit-2.12.1-SNAPSHOT-949.jar:?]
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423) ~[?:?]
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387) ~[?:?]
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312) ~[?:?]
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843) ~[?:?]
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808) ~[?:?]
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188) ~[?:?]
```

### To Reproduce

//snow (a high number)

### Expected behaviour

The out of bounds exception.

### Screenshots / Videos

image

### Error log (if applicable)

https://mclo.gs/0nVcUh7

### Fawe Debugpaste

https://athion.net/ISPaster/paste/view/48232a1c853a424db85191cf316f7f1b

### Fawe Version

FastASyncWorldEdit-949

### Checklist

- [X] I have included a Fawe debugpaste.
- [X] I am using the newest build from https://ci.athion.net/job/FastAsyncWorldEdit/ and the issue still persists.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by dordsor21**
- fixes #2963

### CODE REVIEW

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java [hunk 1]
```diff
@@ -120,20 +120,19 @@ public boolean apply(BlockVector3 position, int depth) throws WorldEditException
...
-            } else if (states.containsKey(stair) && block.getState(stair).equalsIgnoreCase("bottom")) {
+            } else if (type.hasProperty(stair) && block.getState(stair).equalsIgnoreCase("bottom")) {
                 return false;
             } else {
                 return false;
             }
             //FAWE end
-        } else if (!block.getBlockType().id().toLowerCase(Locale.ROOT).contains("ice") && block
-                .getBlockType()
+        } else if (!block.getBlockType().id().toLowerCase(Locale.ROOT).contains("ice") && block.getBlockType()
```
Comment (PierreSchwang) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java:135]: Worth updating maybe, minecraft provides block tags (present in the BlockCategories class as constants) `snow_layer_cannot_survive_on` and `snow_layer_can_survive_on`, which could make this more readable and future-proof (new ice blocks, new spawn logic, etc)

**Review by PierreSchwang — APPROVED** [review]

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java (hunk 1 above)
Comment (dordsor21) [reply worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java:135]: Ah good shout, cheers

File: worldedit-core/src/main/java/com/sk89q/worldedit/world/block/BlockType.java [hunk 2]
```diff
@@ -248,18 +248,17 @@ public boolean hasProperty(PropertyKey key) {
...
-        int ordinal = key.getId();
-        Property<?> property;
+    public boolean hasProperty(Property<?> property) {
+        int ordinal = property.getKey().getId();
+        Property<?> selfProperty;
         return this.settings.propertiesMapArr.length > ordinal
-                && (property = this.settings.propertiesMapArr[ordinal]) != null
-                && property.getClass() == propertyType;
+                && (selfProperty = this.settings.propertiesMapArr[ordinal]) != null
+                && selfProperty == property;
```
Comment (SirYwell) [inline worldedit-core/src/main/java/com/sk89q/worldedit/world/block/BlockType.java:261]: I don't think this is sufficient. For some properties, we create them on the spot (i.e. EnumProperty and DirectionalProperty), so different instances might represent the same actual property. I think using `equals` and overriding it in the relevant property classes is more reliable.

The null check can be removed too I guess, then the assignment isn't needed anymore either.

Comment (dordsor21) [reply worldedit-core/src/main/java/com/sk89q/worldedit/world/block/BlockType.java:261]: Yep, I improved the equals in AbstractProperty and then forgot to use it anyway :)

**Review by SirYwell — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Pull request merged**